### PR TITLE
Fix tensorflow model cross version tests

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Fix the following error:

https://github.com/mlflow/mlflow/actions/runs/4156474523/jobs/7190301167#step:13:557

```
  File "/usr/share/miniconda/envs/mlflow-8618db12fc55d6f8c258c30d9277a6abeec09e35/lib/python3.8/site-packages/tensorflow/core/framework/tensor_shape_pb2.py", line 36, in <module>
    _descriptor.FieldDescriptor(
  File "/usr/share/miniconda/envs/mlflow-8618db12fc55d6f8c258c30d9277a6abeec09e35/lib/python3.8/site-packages/google/protobuf/descriptor.py", line 560, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).
```

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
What caused the issue?

**Comment by WeichenXu123** [comment]
We can confirm that,

now we run `conda env create -f tests/tensorflow/mlflow-128-tf-23-env.yaml`, we will create an environment with `protobuf==4.21.12`, and we know that protobuf 4.21 version is not compatible with Tensorflow model saving.

For several days ago, it should install older version of protobuf when running the same command, I guess conda-forge might upload some new version package that affect the version resolution.

**Comment by harupy** [comment]
I investigated the cause. By running `docker run --rm python:3.8 bash -c "pip install tensorflow==2.3.0 pipdeptree && pipdeptree --reverse --packages protobuf"`, you can get:

```
protobuf==4.21.12
  - tensorboard==2.12.0 [requires: protobuf>=3.19.6]
    - tensorflow==2.3.0 [requires: tensorboard>=2.3.0,<3]
  - tensorflow==2.3.0 [requires: protobuf>=3.9.2]
```

This means `tensorflow` and `tensorboard` require `protobuf`.

---

The latest version of `tensorboard` was released on Feb 10 so it looks suspicious.: https://pypi.org/project/tensorboard/2.12.0

---

This file determines `tensorboard` dependencies:
https://github.com/tensorflow/tensorboard/blob/master/tensorboard/pip_package/requirements.txt

---

I'm checking the commit history:
https://github.com/tensorflow/tensorboard/commits/master/tensorboard/pip_package/requirements.txt

---

Found https://github.com/tensorflow/tensorboard/commit/78f97c28273f3909a69c0b9f36ffcc2d0130ca48

**Comment by harupy** [comment]
> I guess conda-forge might upload some new version package that affect the version resolution.

`tests/tensorflow/mlflow-128-tf-23-env.yaml` doesn't contain conda dependencies. `conda-forge` is unrelated.

**Comment by WeichenXu123** [comment]
> I investigated the cause. By running `docker run --rm python:3.8 bash -c "pip install tensorflow==2.3.0 pipdeptree && pipdeptree --reverse --packages protobuf"`, you can get:
> 
> ```
> protobuf==4.21.12
>   - tensorboard==2.12.0 [requires: protobuf>=3.19.6]
>     - tensorflow==2.3.0 [requires: tensorboard>=2.3.0,<3]
>   - tensorflow==2.3.0 [requires: protobuf>=3.9.2]
> ```
> 
> This means `tensorflow` and `tensorboard` require `protobuf`.
> 
> The latest version of `tensorboard` was released on Feb 10 so it looks suspicious because this issue started happening 3 or 4 days ago: https://pypi.org/project/tensorboard/2.12.0
> 
> This file determines `tensorboard`'s dependencies: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/pip_package/requirements.txt
> 
> I'm checking the commit history of this file: https://github.com/tensorflow/tensorboard/commits/master/tensorboard/pip_package/requirements.txt
> 
> Found [tensorflow/tensorboard@78f97c2](https://github.com/tensorflow/tensorboard/commit/78f97c28273f3909a69c0b9f36ffcc2d0130ca48) changed `protobuf >= 3.9.2, < 4` to `protobuf >= 3.9.16`.

Great investigation!

File: mlflow/tensorflow/__init__.py [hunk 1]
```diff
@@ -844,6 +844,7 @@ def predict(self, data):
             return pandas.DataFrame(self.keras_model.predict(data.values), index=data.index)
 
         supported_input_types = (np.ndarray, list, tuple, dict)
+
```
Comment (harupy) [inline mlflow/tensorflow/__init__.py:847]:
```suggestion
```

**Review by harupy — APPROVED** [review]

**Pull request merged**